### PR TITLE
refactor(registry): Refactor get_changes_since.

### PR DESCRIPTION
**Pull request by daniel-wong-dfinity-org**
In a nut shell, this adds a local `main` function that does the "real work" (including error handling). Whereas, operations that are common to all cases (error, or ok) are done directly in the body of get_changes_since.

These changes were originally part of #1206, but were spun out.

I recommend viewing this using side by side diff is recommended; unified is not so helpful. Also, I recommend enabling disregard whitespace changes.

### CODE REVIEW

File: rs/registry/canister/canister/canister.rs [hunk 1]
```diff
@@ -185,60 +185,67 @@ ic_nervous_system_common_build_metadata::define_get_build_metadata_candid_method
...
+            }
         };
+        let version = request;
 
-        let response: Vec<u8> =
-            serialize_get_changes_since_response(response).expect("Error serializing response");
+        // All requirements met. Proceed with "real work".
 
-        reply(&response);
-        return;
```
Comment (daniel-wong-dfinity-org) [inline rs/registry/canister/canister/canister.rs]: This is duplicated later. Getting rid of this duplication is one of the goals of this PR.

File: rs/registry/canister/canister/canister.rs [hunk 2]
```diff
@@ -185,60 +185,67 @@ ic_nervous_system_common_build_metadata::define_get_build_metadata_candid_method
...
+        if !ALLOWED_CALLER_CLASSES.contains(&class) {
+            return Err((
+                Code::Authorization,
+                format!(
                     "Caller must be self-authenticating, or anonymous (but was {:?}).",
                     class,
                 ),
-                key: vec![],
-            }),
-            ..Default::default()
```
Comment (daniel-wong-dfinity-org) [inline rs/registry/canister/canister/canister.rs:204]: This is useless, and repeated later. Getting rid of this is one of the minor improvements in this PR.

**Review by daniel-wong-dfinity-org — COMMENTED** [review]
Some guidance that reviewers might find helpful.

File: rs/registry/canister/canister/canister.rs [hunk 3]
```diff
@@ -185,60 +185,63 @@ ic_nervous_system_common_build_metadata::define_get_build_metadata_candid_method
...
+        let registry = registry();
 
-            let max_versions = registry
-                .count_fitting_deltas(version, MAX_REGISTRY_DELTAS_SIZE)
-                .min(MAX_VERSIONS_PER_QUERY);
+        let max_versions = registry
+            .count_fitting_deltas(version, MAX_REGISTRY_DELTAS_SIZE)
+            .min(MAX_VERSIONS_PER_QUERY);
+
+        Ok(RegistryGetChangesSinceResponse {
```
Comment (max-dfinity) [inline rs/registry/canister/canister/canister.rs:218]: NCR: Minor thing, but I might have kept a symmetry around the Err/Ok not being the response object itself, and instead create the response object below?  

Random thoughts:
This pattern sort of reminds me of normal web applications with an API layer, where the API layer translates the request into internal operations (ideally in a simple way) and then translates that back into a response.  It's largely what we've been working on doing with our API decoupling.  

It makes me wonder a bit if maybe some of what's happening in this function ought to live somewhere else anyway.

**Review by max-dfinity — APPROVED** [review]

**Pull request merged**